platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/MerkleStateRoot.java

```diff
@@ -524,118 +474,112 @@ private boolean checkNodeIndex(final int index, @NonNull final String label) {
             this.stateMetadata = requireNonNull(stateMetadata);
-            this.stateKeys = Collections.unmodifiableSet(stateMetadata.keySet());
+            this.stateIds = Collections.unmodifiableSet(stateMetadata.keySet());
             this.kvInstances = new HashMap<>();
             this.singletonInstances = new HashMap<>();
             this.queueInstances = new HashMap<>();
         }
 
         @NonNull
         @Override
-        public <K, V> ReadableKVState<K, V> get(@NonNull String stateKey) {
-            final ReadableKVState<K, V> instance = (ReadableKVState<K, V>) kvInstances.get(stateKey);
+        public <K, V> ReadableKVState<K, V> get(final int stateId) {
+            final ReadableKVState<K, V> instance = (ReadableKVState<K, V>) kvInstances.get(stateId);
             if (instance != null) {
                 return instance;
             }
 
-            final var md = stateMetadata.get(stateKey);
+            final var md = stateMetadata.get(stateId);
             if (md == null || md.stateDefinition().singleton()) {
-                throw new IllegalArgumentException("Unknown k/v state key '" + stateKey + ";");
+                throw new IllegalArgumentException("Unknown k/v state key '" + stateId + ";");
             }
 
             final var node = findNode(md);
             if (node instanceof VirtualMap v) {
                 final var ret = createReadableKVState(md, v);
-                kvInstances.put(stateKey, ret);
-                return ret;
-            } else if (node instanceof MerkleMap m) {
-                final var ret = createReadableKVState(md, m);
-                kvInstances.put(stateKey, ret);
+                kvInstances.put(stateId, ret);
                 return ret;
             } else {
                 // This exception should never be thrown. Only if "findNode" found the wrong node!
-                throw new IllegalStateException("Unexpected type for k/v state " + stateKey);
+                throw new IllegalStateException("Unexpected type for k/v state " + stateId);
             }
         }
 
         @NonNull
         @Override
-        public <S> ReadableSingletonState<S> getSingleton(@NonNull String stateKey) {
-            final ReadableSingletonState<S> instance = (ReadableSingletonState<S>) singletonInstances.get(stateKey);
+        public <S> ReadableSingletonState<S> getSingleton(final int stateId) {
+            final ReadableSingletonState<S> instance = (ReadableSingletonState<S>) singletonInstances.get(stateId);
             if (instance != null) {
                 return instance;
             }
 
-            final var md = stateMetadata.get(stateKey);
+            final var md = stateMetadata.get(stateId);
             if (md == null || !md.stateDefinition().singleton()) {
-                throw new IllegalArgumentException("Unknown singleton state key '" + stateKey + "'");
+                throw new IllegalArgumentException("Unknown singleton state key '" + stateId + "'");
```

Comment:
```suggestion
                throw new IllegalArgumentException("Unknown singleton state id '" + stateId + "'");
```

Reply:
A format might be clearer:
```suggestion
                throw new IllegalArgumentException("Unknown singleton state key '%d'".formatted(stateId));
```